### PR TITLE
Add CT::Option

### PR DESCRIPTION
**Pull request by randombit**
GH #4172

### CODE REVIEW

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:182]:
```suggestion
/**
* A CT::Option<T> is either a valid T (Some), or a default T (None)
*/
template <std::default_initializable T>
class Option {
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:188]:
```suggestion
      constexpr Option(T v, Choice valid) : m_is_some(valid), m_value(std::move(v)) {}

      constexpr Option(T v) : m_is_some(Choice::yes()), m_value(std::move(v)) {}

      constexpr Option() : m_is_some(Choice::no()), m_value(T()) {}
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:204]:
```suggestion
      /**
      * Apply a function to the inner value and return a new Option
      * which contains that value. This is constant time only if @p f is.
      *
      * @note The function will always be called, even if the Option is None. It
      *       must be prepared to handle any possible state of T.
      */
      template <std::invocable<const T&> F>
      constexpr auto map(F f) const -> Option<std::invoke_result_t<F, const T&>> {
         return {f(m_value), m_is_some};
      }
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:226]:
```suggestion
      /// Variable time
      std::optional<T> as_optional_vartime() const {
         if(m_is_some.as_bool()) {
            return {m_value};
         } else {
            return {};
         }
      }
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:217]: Probably a good idea to scope this to types that actually provide this method. And also make the copy of `other` visible via the interface (to optimize usage with rvalue refs for instance):

```suggestion
      /// Returns either the inner value or the alternative, in constant time
      T unwrap_or(T other) const 
         requires concepts::ct_conditional_assignable<T>
      {
         other.conditional_assign(m_is_some, m_value);
         return other;
      }
```

... and the concept itself:

```C++
namespace concepts {

template <typename T>
concept ct_conditional_assignable = requires(T lhs, const T& rhs, Choice c) { lhs.conditional_assign(c, rhs); };

}  // namespace concepts

}
```

... if you wanted to, you could even provide a variable time overload if the concept isn't fulfilled. Though, I'd consider this fairly reckless, as it would potentially fool the user if their type doesn't model the concept (perhaps even subtly inadvertantly).

**Review by reneme — COMMENTED** [review]
Nice idea!

Frankly, I would suggest to stay with the STL's nomenclature of things, though. I.e.

* `unwrap()` -> `value()`
* `unwrap_or()` -> `value_or()`
* `is_some()` -> `has_value()`
* `map()` -> `transform()` (C++23: [std::optional<>::transform](https://en.cppreference.com/w/cpp/utility/optional/transform))

File: src/lib/utils/ct_utils.h
Comment (reneme) [reply src/lib/utils/ct_utils.h:204]: It might make sense to look into the other overloads of [`std::optional::transform`](https://en.cppreference.com/w/cpp/utility/optional/transform). For instance, if `this` is an rvalue-ref we could `std::move(*this)` into the function.

Comment (reneme) [reply src/lib/utils/ct_utils.h:204]: Another thing to consider: If `m_is_some` is indicating "None", the resulting `Option<U>` will be "None" but its "value" won't be a default-initialized `U` but whatever came back from `f()`. That may or may not be problematic, but it is probably worth documenting that `m_is_some == None` does not implicate that `m_value` is default inited.

File: src/lib/utils/ct_utils.h
Comment (randombit) [reply src/lib/utils/ct_utils.h:217]: I'd be quite against a variable time overload. That said it would be nice if we could handle this more generically, ie `CT::Option<uint32_t>` is entirely a sensible thing, and it's quite possible to implement `unwrap_or` in constant time as well, but requires additional help. I'll ponder this.

Comment (randombit) [reply src/lib/utils/ct_utils.h]: Nice way of expressing the conditional assignment btw.

I made a change such that `value_or` can handle either types which explicitly provided an appropriate `conditional_assign`, and for unsigned integral types which make use of `CT::Mask` to perform the selection.

File: src/tests/test_ct_utils.cpp
Comment (randombit) [inline src/tests/test_ct_utils.cpp:14]: I moved the CT tests to a new file as `test_utils.cpp` is getting quite large. The Mask and Choice tests are unchanged

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:437]:
```suggestion
/**
* A CT::Option<T> is either a valid T, or not
*
* To maintain constant time behavior a value must always be stored.
* A CT::Choice tracks if the value is valid or not. It is not possible
* to access the inner value if the Choice is unset.
*/
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:493]: Nice!

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:471]: Should that really return a copy of `m_value`? Maybe:

```suggestion
      /// Either returns the value or throws an exception
      const T& value() const {
         BOTAN_STATE_CHECK(m_is_some.as_bool());
         return m_value;
      }
```

... for completeness, [`std::optional` defines this both for lvalue and rvalue `this`](https://en.cppreference.com/w/cpp/utility/optional/value). Not sure we really need that, though. It just optimizes cases where a temporary is passed on.

```C++
takes_an_Option_by_value(returns_an_option_by_value().value())
```

File: src/lib/utils/ct_utils.h
Comment (reneme) [inline src/lib/utils/ct_utils.h:465]: Mhh, here the return type might need some extra case. Imagine `F` returns a reference for some reason, then you would end up instantiating an `Option<T&>` which most certainly isn't what we want.

Either do

```suggestion
      template <std::invocable<const T&> F>
      constexpr auto transform(F f) const -> Option<std::remove_cvref_t<std::invoke_result_t<F, const T&>>> {
         return {f(m_value), m_is_some};
      }
```

and/or exclude references for `typename T` in general with

```C++
template <typename T>
   requires !std::is_reference_t<T>
class Option {
```

**Review by reneme — COMMENTED** [review]
Perhaps also `constexpr` the remaining member functions. I don't see why any of them wouldn't be `constexpr`.

File: src/lib/utils/ct_utils.h
Comment (randombit) [reply src/lib/utils/ct_utils.h:471]: I suppose I was thinking of `Option` as largely for holding types that are cheap to copy (integers, simple arrays, etc) where this wouldn't matter, but yeah no reason not to return a reference here, `CT::Option<vector<uint8_t>>` is entirely sensible and might indeed be something worth using in PK decryption or the like.

File: src/lib/utils/ct_utils.h
Comment (randombit) [reply src/lib/utils/ct_utils.h:465]: `Option<&T>` might make sense in some scenario. I can't think of an immediate use case though

Comment (reneme) [reply src/lib/utils/ct_utils.h]: Perhaps. But this is what `std::reference_wrapper<>` is for, IMHO. Otherwise, imagine this:

```C++
auto foo(Option<MyConstTimeType> o) {
   return o.transform([](const MyConstTimeType& ov) {
      return ov.get_some_member_by_reference();
   });
}
```

... `foo()` would return an `Option<T&>` that would most-certainly be a dangling reference, given that the getter returns a reference to a member of the local copy of `Option<T> o`. That's a source of very subtle memory bugs.

**Review by reneme — APPROVED** [review]

File: src/lib/utils/ct_utils.h
Comment (reneme) [reply src/lib/utils/ct_utils.h]: When we foresee to use it with `std::vector<uint8_t>` it would actually make sense to implement the lvalue/rvalue overloads that `std::optional<>` implements. In fact, this just made me re-think the `strong_type_unwrap()` helper: https://github.com/randombit/botan/pull/4170/files#r1672372281

**Comment by reneme** [comment]
Perhaps its worth to reconsider [this](https://github.com/randombit/botan/pull/4175#discussion_r1672377336) before merging.

File: src/lib/utils/ct_utils.h
Comment (randombit) [reply src/lib/utils/ct_utils.h]: Reading https://stackoverflow.com/questions/24824432/what-is-use-of-the-ref-qualifier-const I am unconvinced by `const &&` and don't really see benefit. I'm going to merge this as is and if/when Option of non-trivally-copyable types comes up we can consider the best course.

**Pull request merged**
